File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/RecyclerFragment.java
package forpdateam.ru.forpda.ui.fragments;

import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import forpdateam.ru.forpda.R;

/**
 * Created by radiationx on 14.08.17.
 */

public abstract class RecyclerFragment extends TabFragment {
    protected SwipeRefreshLayout refreshLayout;
    protected RecyclerView recyclerView;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        baseInflateFragment(inflater, R.layout.fragment_base_list);
        refreshLayout = (SwipeRefreshLayout) findViewById(R.id.swipe_refresh_list);
        recyclerView = (RecyclerView) findViewById(R.id.base_list);
        contentController.setMainRefresh(refreshLayout);
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setListsBackground();
        recyclerView.setHasFixedSize(true);
        refreshLayoutStyle(refreshLayout);
    }

    protected void listScrollTop() {
        new Handler().postDelayed(() -> {
            recyclerView.smoothScrollToPosition(0);
        }, 225);
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/RecyclerFragment.kt
package forpdateam.ru.forpda.ui.fragments

import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.appbar.AppBarLayout
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.databinding.FragmentBaseListBinding
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Created by radiationx on 14.08.17.
 */

abstract class RecyclerFragment : TabFragment(R.layout.fragment_base_list), TabTopScroller {

    private val binding by tabBinding(FragmentBaseListBinding::bind)

    protected val refreshLayout: SwipeRefreshLayout
        get() = binding.swipeRefreshList
    protected val recyclerView: RecyclerView
        get() = binding.baseList

    private var listScrollY = 0
    private var appBarOffset = 0

    private lateinit var topScroller: RecyclerTopScroller

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        contentController.setMainRefresh(refreshLayout)
        setListsBackground()
        recyclerView.setHasFixedSize(true)
        refreshLayoutStyle(refreshLayout)
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(
                recyclerView: RecyclerView,
                dx: Int,
                dy: Int
            ) {
                super.onScrolled(recyclerView, dx, dy)
                listScrollY = recyclerView.computeVerticalScrollOffset()
                updateToolbarShadow()
            }
        })

        appBarLayout.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener { _, i ->
            appBarOffset = i
            updateToolbarShadow()
        })

        topScroller = RecyclerTopScroller(recyclerView, appBarLayout)
    }

    override fun isShadowVisible(): Boolean {
        //Log.e("kololo", "isShadowVisible " + appBarOffset + " " + listScrollY + " -> " + (appBarOffset != 0 || listScrollY > 0));
        return appBarOffset != 0 || listScrollY > 0
    }

    protected fun listScrollTop() {
        GlobalScope.launch(Dispatchers.Main) {
            delay(225)
            recyclerView.smoothScrollToPosition(0)
        }
    }

    override fun toggleScrollTop() {
        topScroller.toggleScrollTop()
    }
}
